dcos/dcos-ui
# DCOS_OSS-4054: Implement endpoint for delayed reset functionality

Implement the endpoint used for resetting
delayed services using the marathon API.

Closes https://jira.mesosphere.com/browse/DCOS_OSS-4054

## Testing
Since this is just the api, we have to make some changes to the code to test it.
1. In `dcos-ui/plugins/services/src/js/components/modals/ServiceModals.js` change line 69 from
`let deleteItem = force => this.props.actions.deleteService(service, force);`
to 
`let deleteItem = force => this.props.actions.resetDelayedService(service, force);`
2. In `dcos-ui/plugins/services/src/js/events/MarathonActions.js`
add `console.log("success")` on line 270 and `console.log("error ", xhr)` on line 275.
3. Create a service that should be running.
Example JSON:
```
{
  "id": "/non-delayed",
  "instances": 1,
  "portDefinitions": [],
  "container": {
    "type": "MESOS",
    "volumes": []
  },
  "cpus": 0.1,
  "mem": 128,
  "requirePorts": false,
  "networks": [],
  "healthChecks": [],
  "fetch": [],
  "constraints": [],
  "cmd": "sleep 1000"
}
```
4. Create a service that will be delayed.
Example JSON from the jira:
```
{
"id": "/delayed",
"backoffFactor": 10,
"backoffSeconds": 1,
"cmd": "exit 1",
"container": {
"type": "MESOS",
"volumes": []
},
"cpus": 0.1,
"disk": 0,
"instances": 10,
"maxLaunchDelaySeconds": 3600,
"mem": 10,
"gpus": 0,
"networks": [
{
"mode": "host"
}
],
"portDefinitions": [],
"requirePorts": false,
"upgradeStrategy": {
"maximumOverCapacity": 1,
"minimumHealthCapacity": 1
},
"killSelection": "YOUNGEST_FIRST",
"unreachableStrategy": {
"inactiveAfterSeconds": 0,
"expungeAfterSeconds": 0
},
"healthChecks": [],
"fetch": [],
"constraints": []
}
```
5. Try to delete (actually reset, we are just highjacking the modal) both of them and observe the console. Trying to reset the delayed one should result in a success, trying to delete the other one should result in an error.
6. Verify the same for pods.
Example JSON:
non-delayed
```
{
  "id": "/non-delayed-pod",
  "containers": [
    {
      "name": "container-1",
      "resources": {
        "cpus": 0.1,
        "mem": 128
      },
      "exec": {
        "command": {
          "shell": "sleep 1000"
        }
      }
    }
  ],
  "scaling": {
    "instances": 1,
    "kind": "fixed"
  },
  "networks": [
    {
      "mode": "host"
    }
  ],
  "volumes": [],
  "fetch": [],
  "scheduling": {
    "placement": {
      "constraints": []
    }
  }
}
```
delayed
```
{
  "id": "/delayed-pod",
  "version": "2019-05-28T11:20:07.349Z",
  "containers": [
    {
      "name": "container-1",
      "resources": {
        "cpus": 0.1,
        "mem": 10,
        "disk": 0,
        "gpus": 0
      },
      "exec": {
        "command": {
          "shell": "exit 1"
        }
      }
    }
  ],
  "networks": [
    {
      "mode": "host"
    }
  ],
  "scaling": {
    "instances": 10,
    "kind": "fixed"
  },
  "scheduling": {
    "backoff": {
      "backoff": 1,
      "backoffFactor": 1.15,
      "maxLaunchDelay": 300
    },
    "upgrade": {
      "minimumHealthCapacity": 1,
      "maximumOverCapacity": 1
    },
    "killSelection": "YOUNGEST_FIRST",
    "unreachableStrategy": {
      "inactiveAfterSeconds": 0,
      "expungeAfterSeconds": 0
    },
    "placement": {
      "constraints": []
    }
  },
  "executorResources": {
    "cpus": 0.1,
    "mem": 32,
    "disk": 10
  },
  "volumes": [],
  "fetch": []
}
```
7. Verify that the added tests make sense.

## Trade-offs
None.

## Dependencies
None.

## Screenshots
![Снимка от 2019-05-29 13-10-27](https://user-images.githubusercontent.com/40791275/58549765-f30a3500-8214-11e9-9e64-41b16344abe7.png)
